cpp/src/gandiva/function_registry_binaryfn.cc

```diff
@@ -0,0 +1,313 @@
+
+    // date_add(date64, int32), date_add(timestamp, int32)
+    BINARY_GENERIC_SAFE_NULL_IF_NULL(date_add, date64, int32, date64),
+    BINARY_GENERIC_SAFE_NULL_IF_NULL(add, date64, int32, date64),
+    BINARY_GENERIC_SAFE_NULL_IF_NULL(date_add, timestamp, int32, timestamp),
+    BINARY_GENERIC_SAFE_NULL_IF_NULL(add, timestamp, int32, timestamp),
+
+    // date_add(date64, int64), date_add(timestamp, int64)
+    BINARY_GENERIC_SAFE_NULL_IF_NULL(date_add, date64, int64, date64),
+    BINARY_GENERIC_SAFE_NULL_IF_NULL(add, date64, int64, date64),
+    BINARY_GENERIC_SAFE_NULL_IF_NULL(date_add, timestamp, int64, timestamp),
+    BINARY_GENERIC_SAFE_NULL_IF_NULL(add, timestamp, int64, timestamp),
+
+    // date_add(int32, date64), date_add(int32, timestamp)
+    BINARY_GENERIC_SAFE_NULL_IF_NULL(date_add, int32, date64, date64),
+    BINARY_GENERIC_SAFE_NULL_IF_NULL(add, int32, date64, date64),
+    BINARY_GENERIC_SAFE_NULL_IF_NULL(date_add, int32, timestamp, timestamp),
+    BINARY_GENERIC_SAFE_NULL_IF_NULL(add, int32, timestamp, timestamp),
+
+    // date_add(int64, date64), date_add(int64, timestamp)
+    BINARY_GENERIC_SAFE_NULL_IF_NULL(date_add, int64, date64, date64),
+    BINARY_GENERIC_SAFE_NULL_IF_NULL(add, int64, date64, date64),
+    BINARY_GENERIC_SAFE_NULL_IF_NULL(date_add, int64, timestamp, timestamp),
+    BINARY_GENERIC_SAFE_NULL_IF_NULL(add, int64, timestamp, timestamp),
+
+    // date_sub(date64, int32), subtract and date_diff
+    BINARY_GENERIC_SAFE_NULL_IF_NULL(date_sub, date64, int32, date64),
+    BINARY_GENERIC_SAFE_NULL_IF_NULL(subtract, date64, int32, date64),
+    BINARY_GENERIC_SAFE_NULL_IF_NULL(date_diff, date64, int32, date64),
+    // date_sub(timestamp, int32), subtract and date_diff
+    BINARY_GENERIC_SAFE_NULL_IF_NULL(date_sub, timestamp, int32, date64),
+    BINARY_GENERIC_SAFE_NULL_IF_NULL(subtract, timestamp, int32, date64),
+    BINARY_GENERIC_SAFE_NULL_IF_NULL(date_diff, timestamp, int32, date64),
+
+    // date_sub(date64, int64), subtract and date_diff
+    BINARY_GENERIC_SAFE_NULL_IF_NULL(date_sub, date64, int64, date64),
+    BINARY_GENERIC_SAFE_NULL_IF_NULL(subtract, date64, int64, date64),
+    BINARY_GENERIC_SAFE_NULL_IF_NULL(date_diff, date64, int64, date64),
+    // date_sub(timestamp, int64), subtract and date_diff
+    BINARY_GENERIC_SAFE_NULL_IF_NULL(date_sub, timestamp, int64, date64),
+    BINARY_GENERIC_SAFE_NULL_IF_NULL(subtract, timestamp, int64, date64),
+    BINARY_GENERIC_SAFE_NULL_IF_NULL(date_diff, timestamp, int64, date64),
+
+
+    BINARY_RELATIONAL_SAFE_NULL_IF_NULL(starts_with, utf8),
+    BINARY_RELATIONAL_SAFE_NULL_IF_NULL(ends_with, utf8)
+  };
+
+  static FunctionRegistryBin::SignatureMap map;
+
+  const int num_entries =
+      static_cast<int>(sizeof(binary_fn_registry_) / sizeof(NativeFunction));
+  for (int i = 0; i < num_entries; i++) {
+    const NativeFunction* entry = &binary_fn_registry_[i];
+
+    DCHECK(map.find(&entry->signature()) == map.end());
```

Comment:
What happens if this function is called twice? It looks like this check may trigger. I'm not sure that's a concern.